### PR TITLE
WIP: Shap integration

### PR DESCRIPTION
**Pull request by alexcoca**
This is a future pull request for the integretion of the KernelShap algorithm into the alibi library.

### CODE REVIEW

**Comment by jklaise** [comment]
Notes on the example notebooks (tried using ReviewNB but didn't work for posting comments):

- [x] examples/kernel_shap_wine_intro.ipynb
  - Potentially confusing use of `cross_val_score`
  - Missing discussion on how missing values are computed using the background dataset
  - Set `l1_reg` to something to suppress deprecation warnings (e.g. `"num_features(10)"`)
  - Mention stacked force-plots y-value for feature is `baseline+shap_value` 

- [ ] examples/kernel_shap_wine_lr.ipynb

- [ ] examples/kernel_shap_adult_lr.ipynb

**Comment by amit1rrr** [comment]
@jklaise Sorry about the ReviewNB posting error. Looks like GitHub post_review API is returning with file path is invalid error for this PR. I am going to try posting a test review, hope you don't mind. I will delete it as soon as testing is done.

File: alibi/explainers/shap.py [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+                cat_vars_enc_dim: Sequence[int] = None,
+                **kwargs) -> Explanation:
+        """
+        Explains the instances in the array X.
+
+        Parameters
+        ----------
+        X
+            Array with instances to be explained.
+        explanation_type
```
Comment (jklaise) [inline alibi/explainers/shap.py]: `explanation_type` shouldn't be a user argument, this is specified as an atribute of the explainer to signal what explanation types are available.

File: alibi/explainers/shap.py [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+                    "Specified both groups as well as summarisation for categorical variables. "
+                    "By grouping, only one shap value is estimated for each categorical variable"
+                    "so summarisation is not necessary."
+                )
+                self.summarise_result = False
+            else:
+                shap_values = sum_categories(shap_values, cat_vars_start_idx, cat_vars_enc_dim)
+        expected_value = self.explainer.expected_value
+
+        self.meta["explnations"][0] = explanation_type
```
Comment (jklaise) [inline alibi/explainers/shap.py]: unnecessary

File: alibi/explainers/tests/test_kernel_shap.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import pytest
+import numpy as np
+
+from alibi.explainers.kernel_shap import sum_categories
+
+
+def get_random_maxtrix(*, n_rows=500, n_cols=100):
```
Comment (jklaise) [inline alibi/explainers/tests/test_kernel_shap.py]: typo

**Comment by jklaise** [comment]
Locked out from ReviewNB for some reason...

Couple of comments on the `shap_adult_lr` notebook:
- Would be nice to add a bit more detail about reading the decision plots
- In various places need to pass in `hyp_record_perm[0]` for the decision plot to run

File: alibi/explainers/tests/test_kernel_shap.py (hunk 3 above)
Comment (alexcoca) [reply alibi/explainers/tests/test_kernel_shap.py]: Not sure I see the typo?

File: alibi/explainers/shap.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+                  **kwargs,
+                  ):
+        """
+        Parameters
+        ----------
+            See constructor.
+        """
+        pass
+
+    @_get_data.register(shap.common.Data)
```
[the remark this answers is not in the document]
Comment (alexcoca) [reply alibi/explainers/shap.py]: Potentially, but we also go over the character limit for some of them if we use the type hint so in the interest of nice formatting and consistency I went for this option.

**Comment by alexcoca** [comment]
<p>All comments now addressed. </p>

**Comment by alexcoca** [comment]
Closing as this is a WIP pull request and is superseded by #191

**Pull request closed without merging**